cli/module_build.go

```diff
@@ -410,10 +408,33 @@ func ReloadModuleAction(c *cli.Context) error {
 	if err != nil {
 		return err
 	}
-	return restartModule(c, vc, partID, manifest)
+	part, err := vc.getRobotPart(partID)
+	if err != nil {
+		return err
+	}
+	needsRestart := true
+	if !c.Bool(moduleBuildRestartOnly) {
+		if !c.Bool(moduleBuildFlagNoBuild) {
+			if manifest == nil {
+				return fmt.Errorf(`manifest not found at "%s". manifest required for build`, moduleFlagPath)
+			}
+			err = moduleBuildLocalAction(c, manifest)
+			if err != nil {
+				return err
+			}
+		}
+		needsRestart, err = configureModule(vc, manifest, part.Part)
+		if err != nil {
+			return err
+		}
+	}
+	if needsRestart {
+		return restartModule(c, vc, part.Part, manifest)
```

Comment:
I think there is a race condition here that I don't think has been addressed elsewhere (might be wrong):

Suppose I have a config with a registry module
`configureModule` updates my config to a local module at `t==0`
`restartModule` is called at `t==1`, triggering the registry module to restart
robot pulls its config at `t==5` (every 5 seconds) and notices a diff and restarts the module once again.

I don't think this will break anything... but it might be surprising to users who are going to be paying a lot of attention to their modules & their robot

Reply:
I was not thinking about this! good catch

couple of cases:
- **within a run** of this command:
  the module is either reconfigured *or* restarted but not both
- this command is called twice **simultaneously**:
  both will pull old config, both will push identical new config
- this command is called twice in **short succession**:
  this is the case you outlined I think. first call pulls old config + writes back changes, second call restarts, but robot has old config so the module remains a registry module, then robot sees new config and switches to local exec module.

zooming out, this happens because there's no way to serialize re**config** and re**start** events. they have different channels and the former takes up to 5 seconds to arrive.

ideal architectural fix: make the reconfig channel low-latency, and use the same channel to deliver restart events (i.e. 'restart counter' on module config or something). this is a project:tm: though.

a potential targeted fix: try to pull config before restarting?

Reply:
> a potential targeted fix: pull config (w/ short timeout) before restarting?

Would decrease the odds of the race occurring significantly (but also make it harder to reproduce and debug once people start reporting the error... tragedy of partially fixing a race condition)

Reply:
SGTM for now as long as it has a comment explaining that a race still exists

Reply:
Oh -- ignoring the case where the user runs this twice (in parallel or quickly) (their fault / undefined behavior imo)

I think you could fix this by adding a new mutex to local robot that prevents config polling & reload module from running at the same time and then adding a config-poll to reload module... but actually, that is awful and fragile. Maybe deferring on this is okay until users are actually using this

Reply:
e43ef4f8 added comment for this but agree no action for now. best improvement would be to reduce latency of config changes (general improvement)

Reply:
Thanks for catching that, @zaporter-work . I think the comment you've put in is fine @abe-winter 👍🏻 . If that extra restart ends up being common/users complain about it, netcode can find a way to ensure it doesn't happen with some extra layer of locking/counting (would require some thought and I think it's not critical to handle in this PR).